pkg/urbit/jets/e/lms.c

```diff
@@ -0,0 +1,2134 @@
+
+    if ((ia_ < 1) || (ib_ > m_u) || (ja_ < 1) || (jb_ > n_u) || (ia_ > ib_) || (ja > jb_))
+    {
+      return u3m_bail(c3__exit);
+    }
+
+    //  Flip the order around so it feels natural.
+    //  This isn't formally necessary, but it's a lot easier to think about.
+    c3_w* u__  = (c3_w*)u3a_malloc((m_u*n_u+1)*sizeof(float32_t));
+    u3r_words(0, m_u*n_u+1, u__, u);
+    c3_w* u_  = (c3_w*)u3a_malloc((m_u*n_u+1)*sizeof(float32_t));
+    u_[m_u*n_u] = u__[m_u*n_u];
+    c3_w iii;
+    for ( iii = 0; iii < m_u*n_u; iii++ )
+    {
+      u_[m_u*n_u-1-iii] = u__[iii];
+    }
+    u3a_free(u__);
+
+    c3_w m_w, n_w;
+    m_w = ib_ - ia_ + 1;
+    n_w = jb_ - ja_ + 1;
+    c3_w* w_  = (c3_w*)u3a_malloc((m_w*n_w+1)*sizeof(float32_t));
+    c3_w p_w = u3qelms_cantor(m_w, n_w);
+    w_[m_w*n_w] = p_w;
+    c3_w ii, jj, mi, mj;
+    for ( ii = ia_-1, mi = 0; ii <= ib_-1; ii++, mi++ )
+    {
+      for ( jj = ja_-1, mj = 0; jj <= jb_-1; jj++, mj++ )
+      {
+        w_[m_w*n_w-(mi*n_w+mj)-1] = u_[ii*n_u+jj];
+      }
+    }
+    u3_noun w = u3i_words(m_w*n_w+1, w_);
+    u3a_free(w_);
+
+    return w;
+  }
+
+  u3_noun
+  u3welms_submatrix(u3_noun cor)
+  {
+    u3_noun u, ia, ib, ja, jb;
+
+    if ( c3n == u3r_mean(cor, u3x_sam_2, &u, u3x_sam_6, &ia, u3x_sam_14, &ib, u3x_sam_30, &ja, u3x_sam_31, &jb, 0) ||
+         c3n == u3ud(u) ||
+         c3n == u3ud(ia) ||
+         c3n == u3ud(ib) ||
+         c3n == u3ud(ja) ||
+         c3n == u3ud(jb) )
+    {
+      return u3m_bail(c3__exit);
+    }
+    else {
+      return u3qelms_submatrix(u, ia, ib, ja, jb);
+    }
+  }
+
+/* augment matrix
+*/
+  u3_noun
+  u3qelms_augment(u3_atom u)  /* @lms */
+  {
+    c3_w mnt_ = u3r_met(3,u)/4;  // mnt_ is the vector length
+
+    union trip p_;
+    p_.c = u3r_word(mnt_, u);
+    u3_atom p__ = u3i_words(1,&(p_.c));
+
+    u3_noun mn_ = u3qelms_decantor(p__);
+    u3_noun m, n, o;
+    u3r_trel(mn_, &m, &n, &o);
+    c3_w m_u, n_u, m_w, n_w;
+    m_u = u3r_word(0, m);  // XXX unnecessarily formal to unwrap like this
+    n_u = u3r_word(0, n);
+    m_w = m_u;
+    n_w = 2*n_u;
+
+    u3_atom v = u3qelms_id(m_u, n_u);
+
+    c3_w* w_  = (c3_w*)u3a_malloc((m_w*n_w+1)*sizeof(float32_t));
+    c3_w p_w = u3qelms_cantor(m_w, n_w);
+    w_[m_w*n_w] = p_w;
+    c3_w ii, jj;
+    for ( ii = 0; ii < m_u; ii++ )
+    {
+      for ( jj = 0; jj < n_u; jj++ )
+      {
+        w_[ii*n_w+n_u+jj] = u3r_word(ii*n_u+jj, u);
+      }
+      for ( jj = 0; jj < n_u; jj++ )
+      {
+        w_[ii*n_w+jj] = u3r_word(ii*n_u+jj, v);
+      }
+    }
+    u3_noun w = u3i_words(m_w*n_w+1, w_);
+    u3a_free(w_);
+
+    return w;
+  }
+
+  u3_noun
+  u3welms_augment(u3_noun cor)
+  {
+    u3_noun u;
+
+    if ( c3n == (u = u3r_at(u3x_sam, cor)) ||
+         c3n == u3ud(u) )
+    {
+      return u3m_bail(c3__exit);
+    }
+    else {
+      return u3qelms_augment(u);
+    }
+  }
+
+/* invert matrix
+   XXX this should become a wrapper not an implementation
```

Comment:
wrapper to sth like BLAS?

Reply:
Right now I've avoided introducing any external dependencies besides SoftFloat.  I looked into doing more but I think the platform and installation situation starts to get more complicated and I'd like more dev weigh-in before taking the step of increasing the number of dependencies.

Reply:
What I think makes the most sense is to have a base matrix/vector library (this one), then a scientific algorithms library (everything past inversion).  Once you get past inversion (eigenvalues and so forth), there are a lot of interdependencies introduced very quickly and the size will balloon.  So these basic data structures become standard, but the scientific/analytical stuff only gets pulled in when necessary.

Reply:
Yeah makes sense, this tbc was a "what does the comment mean" comment not a "so do that" comment :P